File: sdk/src/main/java/org/p2p/solanaj/kits/TokenSwap.java
package org.p2p.solanaj.kits;

import android.util.Log;

import org.p2p.solanaj.core.Account;
import org.p2p.solanaj.core.PublicKey;
import org.p2p.solanaj.core.TransactionInstruction;
import org.p2p.solanaj.core.TransactionRequest;
import org.p2p.solanaj.programs.SystemProgram;
import org.p2p.solanaj.programs.TokenProgram;
import org.p2p.solanaj.programs.TokenSwapProgram;
import org.p2p.solanaj.rpc.RpcClient;
import org.p2p.solanaj.rpc.RpcException;
import org.p2p.solanaj.rpc.types.TokenAccountBalance;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Collections;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class TokenSwap {

    private final RpcClient client;

    public TokenSwap(RpcClient client) {
        this.client = client;
    }

    public String swap(
            Account owner,
            @NonNull Pool.PoolInfo pool,
            Double slippage,
            BigInteger amountIn,
            TokenAccountBalance balanceA,
            TokenAccountBalance balanceB,
            PublicKey wrappedSolAccount,
            @Nullable PublicKey accountAddressA,
            @Nullable PublicKey accountAddressB
    ) throws RpcException {

        Log.d("###", " a " + accountAddressA.toBase58() + " b " + accountAddressB.toBase58());

        final ArrayList<Account> signers = new ArrayList<>(Collections.singletonList(owner));

        TransactionRequest transaction = new TransactionRequest();

        // swap type
        PublicKey source = pool.getTokenAccountA();
        PublicKey tokenA = source.equals(pool.getTokenAccountA()) ? pool.getTokenAccountA() : pool.getTokenAccountB();

        boolean isTokenAEqTokenAccountA = tokenA.equals(pool.getTokenAccountA());

        PublicKey tokenB = isTokenAEqTokenAccountA ? pool.getTokenAccountB() : pool.getTokenAccountA();
        PublicKey mintA = isTokenAEqTokenAccountA ? pool.getMintA() : pool.getMintB();
        PublicKey mintB = isTokenAEqTokenAccountA ? pool.getMintB() : pool.getMintA();

        PublicKey addressA = new PublicKey("QqCCvshxtqMAL2CVALqiJB7uEeE5mjSPsseQdDzsRUo");
        PublicKey addressB = new PublicKey("QqCCvshxtqMAL2CVALqiJB7uEeE5mjSPsseQdDzsRUo");
//        if (isTokenAEqTokenAccountA) {
//            addressA = accountAddressA;
//            addressB = accountAddressB;
//        } else {
//            addressA = accountAddressB;
//            addressB = accountAddressA;
//        }

        TokenProgram.AccountInfoData tokenAInfo = TokenTransaction.getAccountInfoData(client, tokenA, TokenProgram.PROGRAM_ID);
        int space = TokenProgram.AccountInfoData.ACCOUNT_INFO_DATA_LENGTH;
        long balanceNeeded = client.getApi().getMinimumBalanceForRentExemption(space);

        PublicKey fromAccount;
        if (tokenAInfo.isNative()) {
            Account newAccount = new Account();
            PublicKey newAccountPubKey = newAccount.getPublicKey();

            TransactionInstruction createAccountInstruction = SystemProgram.createAccount(
                    owner.getPublicKey(),
                    newAccountPubKey,
                    amountIn.longValue() + balanceNeeded,
                    TokenProgram.AccountInfoData.ACCOUNT_INFO_DATA_LENGTH,
                    TokenProgram.PROGRAM_ID
            );

            TransactionInstruction initializeAccountInstruction = TokenProgram.initializeAccountInstruction(
                    TokenProgram.PROGRAM_ID,
                    newAccountPubKey,
                    wrappedSolAccount,
                    owner.getPublicKey()
            );

            transaction.addInstruction(createAccountInstruction);
            transaction.addInstruction(initializeAccountInstruction);

            signers.add(newAccount);

            fromAccount = newAccountPubKey;
        } else {
            fromAccount = addressA;
        }

        PublicKey toAccount = addressB;
        boolean isWrappedSol = mintB.equals(wrappedSolAccount);

        if (toAccount == null) {
            Account newAccount = new Account();
            PublicKey newAccountPubKey = newAccount.getPublicKey();

            TransactionInstruction createAccountInstruction = SystemProgram.createAccount(
                    owner.getPublicKey(),
                    newAccountPubKey,
                    balanceNeeded,
                    TokenProgram.AccountInfoData.ACCOUNT_INFO_DATA_LENGTH,
                    TokenProgram.PROGRAM_ID
            );

            TransactionInstruction initializeAccountInstruction = TokenProgram.initializeAccountInstruction(
                    TokenProgram.PROGRAM_ID,
                    newAccountPubKey,
                    mintB,
                    owner.getPublicKey()
            );

            transaction.addInstruction(createAccountInstruction);
            transaction.addInstruction(initializeAccountInstruction);

            signers.add(newAccount);

            toAccount = newAccountPubKey;
        }

        TransactionInstruction approve = TokenProgram.approveInstruction(
                TokenProgram.PROGRAM_ID,
                fromAccount,
                pool.getAuthority(),
                owner.getPublicKey(),
                amountIn
        );

        final BigInteger estimatedAmount = TokenSwap.calculateSwapEstimatedAmount(
                balanceA,
                balanceB,
                amountIn
        );
        final BigInteger minimumAmountOut = TokenSwap.calculateSwapMinimumReceiveAmount(
                estimatedAmount,
                slippage
        );

        Log.d("###", " from " + fromAccount.toBase58() + " to " + toAccount.toBase58());

        TransactionInstruction swap = TokenSwapProgram.swapInstruction(
                pool.getAddress(),
                pool.getAuthority(),
                fromAccount,
                tokenA,
                tokenB,
                toAccount,
                pool.getTokenPool(),
                pool.getFeeAccount(),
                pool.getFeeAccount(),
                TokenProgram.PROGRAM_ID,
                pool.getSwapProgramId(),
                amountIn,
                minimumAmountOut
        );

        transaction.addInstruction(approve);
        transaction.addInstruction(swap);

        boolean isNeedCloseAccount = tokenAInfo.isNative() || isWrappedSol;
        PublicKey closeAccountPublicKey = null;
        if (tokenAInfo.isNative()) {
            closeAccountPublicKey = fromAccount;
        } else if (isWrappedSol) {
            closeAccountPublicKey = toAccount;
        }

        if (isNeedCloseAccount && closeAccountPublicKey != null) {
            TransactionInstruction closeAccountInstruction = TokenProgram.closeAccountInstruction(
                    TokenProgram.PROGRAM_ID,
                    closeAccountPublicKey,
                    owner.getPublicKey(),
                    owner.getPublicKey()
            );
            transaction.addInstruction(closeAccountInstruction);
        }

        return client.getApi().sendTransaction(transaction, signers);
    }

    public static BigInteger calculateSwapMinimumReceiveAmount(BigInteger estimatedAmount, double slippage) {
        return BigDecimal.valueOf(estimatedAmount.doubleValue() * (1 - slippage)).toBigInteger();
    }

    public static BigInteger calculateSwapEstimatedAmount(TokenAccountBalance tokenABalance, TokenAccountBalance tokenBBalance, BigInteger inputAmount) {
        return tokenBBalance.getAmount().multiply(inputAmount).divide(tokenABalance.getAmount().add(inputAmount));
    }
}

File: sdk/src/main/java/org/p2p/solanaj/kits/TokenSwap.kt
package org.p2p.solanaj.kits

import org.p2p.solanaj.kits.Pool.PoolInfo
import org.p2p.solanaj.model.core.Account
import org.p2p.solanaj.model.core.PublicKey
import org.p2p.solanaj.model.core.TransactionRequest
import org.p2p.solanaj.model.types.AccountInfo
import org.p2p.solanaj.model.types.TokenAccountBalance
import org.p2p.solanaj.programs.SystemProgram
import org.p2p.solanaj.programs.TokenProgram
import org.p2p.solanaj.programs.TokenProgram.AccountInfoData
import org.p2p.solanaj.programs.TokenSwapProgram
import org.p2p.solanaj.rpc.RpcException
import java.math.BigDecimal
import java.math.BigInteger
import java.util.ArrayList

class TokenSwap {

    @Throws(RpcException::class)
    suspend fun swap(
        owner: Account,
        pool: PoolInfo,
        slippage: Double,
        amountIn: BigInteger,
        balanceA: TokenAccountBalance,
        balanceB: TokenAccountBalance,
        wrappedSolAccount: PublicKey,
        accountAddressA: PublicKey?,
        accountAddressB: PublicKey?,
        getAccountInfo: suspend (PublicKey) -> AccountInfo,
        getBalanceNeeded: suspend (Long) -> Long,
        sendTransaction: suspend (transaction: TransactionRequest, signers: List<Account>) -> String
    ): String {
        val signers = ArrayList(listOf(owner))
        val transaction = TransactionRequest()

        // swap type
        val source = pool.tokenAccountA
        val tokenA = if (source.equals(pool.tokenAccountA)) pool.tokenAccountA else pool.tokenAccountB
        val isTokenAEqTokenAccountA = tokenA.equals(pool.tokenAccountA)
        val tokenB = if (isTokenAEqTokenAccountA) pool.tokenAccountB else pool.tokenAccountA
        val mintA = if (isTokenAEqTokenAccountA) pool.mintA else pool.mintB
        val mintB = if (isTokenAEqTokenAccountA) pool.mintB else pool.mintA

        val accountInfo = getAccountInfo.invoke(tokenA)
        val tokenAInfo = TokenTransaction.getAccountInfoData(accountInfo, TokenProgram.PROGRAM_ID)
        val space = AccountInfoData.ACCOUNT_INFO_DATA_LENGTH
        val balanceNeeded = getBalanceNeeded.invoke(space.toLong())

        val fromAccount: PublicKey? = if (tokenAInfo.isNative) {
            val newAccount = Account()
            val newAccountPubKey = newAccount.publicKey
            val createAccountInstruction = SystemProgram.createAccount(
                owner.publicKey,
                newAccountPubKey,
                amountIn.toLong() + balanceNeeded,
                AccountInfoData.ACCOUNT_INFO_DATA_LENGTH.toLong(),
                TokenProgram.PROGRAM_ID
            )
            val initializeAccountInstruction = TokenProgram.initializeAccountInstruction(
                TokenProgram.PROGRAM_ID,
                newAccountPubKey,
                wrappedSolAccount,
                owner.publicKey
            )
            transaction.addInstruction(createAccountInstruction)
            transaction.addInstruction(initializeAccountInstruction)
            signers.add(newAccount)
            newAccountPubKey
        } else {
            accountAddressA
        }
        var toAccount = accountAddressB
        val isWrappedSol = mintB.equals(wrappedSolAccount)
        if (toAccount == null) {
            val newAccount = Account()
            val newAccountPubKey = newAccount.publicKey
            val createAccountInstruction = SystemProgram.createAccount(
                owner.publicKey,
                newAccountPubKey,
                balanceNeeded,
                AccountInfoData.ACCOUNT_INFO_DATA_LENGTH.toLong(),
                TokenProgram.PROGRAM_ID
            )
            val initializeAccountInstruction = TokenProgram.initializeAccountInstruction(
                TokenProgram.PROGRAM_ID,
                newAccountPubKey,
                mintB,
                owner.publicKey
            )
            transaction.addInstruction(createAccountInstruction)
            transaction.addInstruction(initializeAccountInstruction)
            signers.add(newAccount)
            toAccount = newAccountPubKey
        }
        val userTransferAuthority = Account()
        val approve = TokenProgram.approveInstruction(
            TokenProgram.PROGRAM_ID,
            fromAccount,
            userTransferAuthority.publicKey,
            owner.publicKey,
            amountIn
        )
        val estimatedAmount = calculateSwapEstimatedAmount(
            balanceA,
            balanceB,
            amountIn
        )
        val minimumAmountOut = calculateSwapMinimumReceiveAmount(
            estimatedAmount,
            slippage
        )
        val swap = TokenSwapProgram.swapInstruction(
            pool.address,
            pool.authority,
            userTransferAuthority.publicKey,
            fromAccount,
            tokenA,
            tokenB,
            toAccount,
            pool.tokenPool,
            pool.feeAccount,
            pool.feeAccount,
            TokenProgram.PROGRAM_ID,
            pool.swapProgramId,
            amountIn,
            minimumAmountOut
        )
        transaction.addInstruction(approve)
        transaction.addInstruction(swap)
        val isNeedCloseAccount = tokenAInfo.isNative || isWrappedSol
        var closeAccountPublicKey: PublicKey? = null
        if (tokenAInfo.isNative) {
            closeAccountPublicKey = fromAccount
        } else if (isWrappedSol) {
            closeAccountPublicKey = toAccount
        }
        if (isNeedCloseAccount && closeAccountPublicKey != null) {
            val closeAccountInstruction = TokenProgram.closeAccountInstruction(
                TokenProgram.PROGRAM_ID,
                closeAccountPublicKey,
                owner.publicKey,
                owner.publicKey
            )
            transaction.addInstruction(closeAccountInstruction)
        }
        signers.add(userTransferAuthority)
        return sendTransaction.invoke(transaction, signers)
    }

    companion object {
        fun calculateSwapMinimumReceiveAmount(estimatedAmount: BigInteger, slippage: Double): BigInteger {
            return BigDecimal.valueOf(estimatedAmount.toDouble() * (1 - slippage)).toBigInteger()
        }

        fun calculateSwapEstimatedAmount(
            tokenABalance: TokenAccountBalance,
            tokenBBalance: TokenAccountBalance,
            inputAmount: BigInteger
        ): BigInteger {
            return tokenBBalance.amount.multiply(inputAmount).divide(tokenABalance.amount.add(inputAmount))
        }
    }
}